libraries/chain/sync_call_context.cpp

```diff
@@ -0,0 +1,21 @@
+#include <eosio/chain/sync_call_context.hpp>
+
+namespace eosio::chain {
+
+sync_call_context::sync_call_context(account_name sender, account_name receiver, uint64_t flags, std::span<const char>data)
+   : sender(sender)
+   , receiver(receiver)
+   , flags(flags)
+   , data(data)
+{
+}
+
+bool sync_call_context::is_read_only()const {
+   return flags & 0b1; // LSB. bit index 0
+}
+
+bool sync_call_context::no_op_if_receiver_not_support_sync_call()const {
+   return flags & 0b10;  // second bit from LSB. bit index 1
```

Comment:
Is there some pattern we can use to get rid of the naked bit constants here and
https://github.com/AntelopeIO/spring/blob/cdb1bebefacb07b4b7bed9d9d9547c4d519928e0/libraries/chain/apply_context.cpp#L262

Simplest being something like
```c++
enum sync_call_flags {
   read_only = 1<<0;
   noop.. = 1<<1;
   last = noop..;
}
```
then flags are valid if `flags <= last`.

Lots of other potential approaches too.

Reply:
Thanks!

https://github.com/AntelopeIO/spring/pull/1266/commits/287e1c585382223ef97edf44ea5d829be8c13c42

Reply:
If we are getting rid of the naked bit constants, I would recommend you do it like this:
https://github.com/AntelopeIO/spring/blob/e186a90a6215405eee3c85a24e9e4358560c0eee/libraries/chain/include/eosio/chain/account_object.hpp#L67-L72

Reply:
No preference from me; using `has_field()` seems, if I'm understanding right, to offer some minor additional paranoia that is good.

My main disappointment is that there is no automatic way to get the last item, thus requiring us to manually specify it. My hunch is it's possible to construct something fancy with constexpr to resolve that but not something I'm advocating we do now.

Reply:
I will change to use `has_field()` so it is more consistent with existing code style.

Reply:
Looked at this further. We need to verify bits other than the first bit can have a value. The `has_field` does not provide a way to do that. Looks like the existing implementation is a simple one.